Implement Network Connect `ClusterIPServiceNetwork` Connectivity

tssurya opened the pull request:
Depends on https://github.com/ovn-kubernetes/ovn-kubernetes/pull/5541

## 📑 Description
<!-- Add a brief description of the pr -->

This PR adds support for partial service connectivity for NetworkConnect. See https://ovn-kubernetes.io/okeps/okep-5224-connecting-udns/okep-5224-connecting-udns/#services for design details.


Fixes #5581 

## Additional Information for reviewers

TBD:

1. Figure out if we really need `ct.new` match in the ACL. Bug is still reproducible, have opened https://issues.redhat.com/browse/FDP-3124/https://issues.redhat.com/browse/FDP-3131 will work with @dceara and OVN team on `pass-related`. Meanwhile we stick with the workaround. See https://cloud-native.slack.com/archives/C08452HR8V6/p1770720413427739. We decided to go the route of doing two Pass ACLs.. see the slack thread for more information. We will FUP on this piece, tracked via https://github.com/ovn-kubernetes/ovn-kubernetes/issues/6048
2. ~Figure out how to re-design "to-self" block that's happening now - that's why CI is red, A few options:~
    ~2.1 Add a higher priority ACL to allow "to-self" subnets on each network's switch? - that's going to be an extra ACL per network - We will go with this - add a higher priority ACL at 5000 to allow src=dst=same network~ THIS IS DONE
    ~2.2 Do 2.1 globally - do we care if we have ACLs of other connected networks in the given network?~
    ~2.3 Any OVN specific hair pin options?~
3. ~The first commit which is a bug fix is not handling upgrades scenario - we probably at this point need to update the names of the load balancers at startup.~
     * ~So looks like `EnsureLBs` is handling this correctly already. It takes `existingCacheLBs` and `LBs` desired.~
     * ~`toDelete` map initializes it with all existing cached LBs.~
     * ~`toDelete[lb.UUID] = lb` -- every existing LB starts as "to delete"~
     * ~`existingLB := existingByName[lb.Name]` -- matches by name since that's the index in DB~
     * ~`delete(toDelete, existingLB.UUID)` -- only removes from toDelete if name matches~
     * ~Everything still in toDelete gets deleted~
     * ~So: old per-node LB (name `_node_router+switch`) stays in `toDelete` because new desired LB has name `_cluster` (no name match). Old LB gets deleted~
     * I'll test and confirm this as well. THIS IS DONE, I checked it. We don't need to do anything specific for this.
4. ~Add UTs for commits 3&4~ THIS IS DONE
5. ~Check with @dceara if load balancer delete will cleanup references or not in OVN? - Ensure a UT exists for this~ - See https://github.com/ovn-kubernetes/ovn-kubernetes/pull/5828#issuecomment-3818734261 , we are good here, THIS IS DONE
6. ~Anything to consider during restart scenario? cache population or sync required?~ THIS IS DONE, Added a repair function
7. ~Reuse existing code for `getNetworkOwnerKeyForNamespace` - see if we can get rid of that util~ - I didn't see any room for reuse, this is DONE
8. **This PR doesn't consider Dynamic UDNs feature - that will need another change on top once this lands.** Tracked via https://github.com/ovn-kubernetes/ovn-kubernetes/issues/5963. A basic level of implementation consideration was done during create case but its not being elaborately fleshed out since dynamic udns landed much after CNC
9. self-note: double-check for error returns properly
10. self-note: https://github.com/ovn-kubernetes/ovn-kubernetes/pull/5828#discussion_r2730888631
11. Discuss with @npinaeva and @trozet about CIP/EIP/LBVIP point - but after testing it out - perhaps now is the time to rename `ClusterIPServiceNetwork` to `ServiceNetwork` OR we stick with the same name but internal behaviour changes a bit. See https://cloud-native.slack.com/archives/C08452HR8V6/p1770802433305249. Will do this in a new PR, opened a tracker for this https://github.com/ovn-kubernetes/ovn-kubernetes/issues/5960


## ✅ Checks
<!-- Make sure your pr passes the CI checks and do check the following fields as needed - -->
- [x] My code requires changes to the documentation
- [ ] if so, I have updated the documentation as required
- [x] My code requires tests
- [x] if so, I have added and/or updated the tests as required
- [x] All the tests have passed in the CI <!-- If not leave a comment as to why the CI is red and if you need help understanding what's wrong -->

## How to verify it
<!--
Did you include unit tests? or end-to-end tests?
How can I manually verify that this patch achieves its objective
-->


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * Per-CNC service LoadBalancerGroup support, service-aware reconciliation, and startup repair for stale CNC-related objects.
  * EndpointSlice mirroring with optional IP transform and new helpers for load‑balancer group ↔ switch associations.
  * New ACL priorities and naming for network-connect service flows.

* **Bug Fixes**
  * Host-endpoint detection and LB external IDs now respect per-network subnet context.
  * Network cleanup removes orphaned load balancers.

* **Tests**
  * Large expansion of unit, integration, and e2e tests for service connectivity, UDN, repair, and cleanup.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

tssurya commented on go-controller/pkg/ovn/controller/networkconnect/topology.go:
```diff
@@ -963,3 +1000,611 @@ func (c *Controller) getNodeSubnet(netInfo util.NetInfo, nodeName string) ([]*ne
+}
+
+// findSwitchesWithCNCACLs returns the names of switches that currently have this CNC's ACLs.
+func (c *Controller) findSwitchesWithCNCACLs(cncName string) (sets.Set[string], error) {
```
I think the reason I didn't use that was because we update the caches at the end of syncNetworkConnections for whichever networks succeeded there..
let me see if I can reuse that but then update the cache only when both `syncNetworkConnections` and `syncPartialConnectivity` both complete

tssurya replied:
I am going to return successful networks from syncPartialConnectivity && syncNetworkConnections - only the intersection of that will be entered into the cache as successfully connected networks..
there is a small downside to doing this which is there could be a case for a network where syncPartialConnectivity succeeded but syncNetworkConnections failed, so we retry and syncPartialConnectivity would also get recomputed.. but I think that is OK ?

npinaeva replied:
yeah it is more difficult to store successfully added networks if addition consists of multiple parts, maybe we can just add another set to track successfully partially connected networks? then they can be independent?

tssurya replied:
I think I've succeeded in removing this util and merging things better - and reusing the connectedNetworks cache instead of building it again from predicates... but let's see what the next round of push tells :D 

resolving this comment

tssurya commented on go-controller/pkg/ovn/controller/networkconnect/topology.go:
```diff
@@ -963,3 +1000,611 @@ func (c *Controller) getNodeSubnet(netInfo util.NetInfo, nodeName string) ([]*ne
+	var dropMatch string
+	switch {
+	case config.IPv4Mode && config.IPv6Mode:
+		dropMatch = fmt.Sprintf("(ip4.src == $%s && ip4.dst == $%s || ip6.src == $%s && ip6.dst == $%s) && ct.new",
```
hmm on second thoughts - the only thing we know for sure is that we don't want traffic from a connected network going to the other connected network for pod-network traffic.. I think let's stick with current match.
for traffic originating from other sources... problem is there could be network policies applied on namespace level and we don't want to drop that intentionally...since the policies are all applied to the common switch and not to transit router of the CNC I think what we have today is accurate for now

npinaeva replied:
yeah this one is not so important, but I just want to clarify the idea here. So for service-only connectivity we don't want networks to be able to talk to each other, but the actual connecting routes still exist in the CNC router, so if the traffic is not dropped at the entrance (aka on the switch)  it will reach another network. That what made me think the this switch should never send traffic to the connected network destination (I would expect users that only request service connectivity to not want anyone to reach their podIPs). I don't fully understand the network policy case: so if the namespace has network policy allowing connected network we will override it with drop, but I would also expect the connected network to not be reachable anyway?

Side-note: we allow 2 networks to be selected by different CNCs as long as they have different connect subnets, but if one CNC has podConnectivity and another doesn't then the traffic will be dropped on the switch by the opartial connectivity ACLs anyway, effectively only allowing serviceConnectivity? Do we need another check for this in the CNC validation?

npinaeva commented on go-controller/pkg/ovn/controller/networkconnect/topology.go:
```diff
@@ -378,6 +435,38 @@ func (c *Controller) syncNetworkConnections(cnc *networkconnectv1.ClusterNetwork
+		// directions: remove this network's LBs from remaining switches AND remove
+		// remaining networks' LBs from this network's switch.
+		if serviceConnectivityDesired {
+			remainingNetworks := cncState.connectedNetworks.Clone()
```
should we call it previouslyConnectedNetworks? remaining made me think it was a set of still-connected networks, which would mean the cleanup for a set of deleted network won't be complete (same for the parameter name in the `cleanupLoadBalancersOps`)

tssurya replied:
I think we ended up simplifying the code when we use LBG now.. so this should get outdated.. I can squash commits in the end if that helps..but keeping the LBG commit separate on top for now for easier review/bugfix capture

tssurya commented:
There redness in CI is because the e2e test fails with the current new update:
There is a potential race when service is intact, but we are deleting and recreating the UDN+pods in the namespace - so service update doesn't happen, the load balancer is not cleaned up when old service controller is stopped... and then when new service controller comes up it deletes this stale LB that causes confusion in CNC cause it had just attached that old LB thinking that's the LB. - I am fixing this by adding an expected count when fetching LBs to ensure we are retrying if we don't find LBs...

tssurya commented on go-controller/pkg/ovn/controller/networkconnect/controller.go:
```diff
@@ -332,6 +392,147 @@ func (c *Controller) reconcileNode(key string) error {
+
+// getNetworkOwnerKeyForNamespace returns the owner key (e.g., "layer3_5") for the primary network
+// of the given namespace. Returns empty string if namespace uses default network.
+// TODO(tssurya): Make this a common utility function and reuse the one in clustermanager/networkconnect/controller.go
```
yea there is a TODO in PR describe so that I don't forget.. just waiting for main code to settle a bit

tssurya commented on go-controller/pkg/ovn/controller/networkconnect/topology.go:
```diff
@@ -963,3 +1000,611 @@ func (c *Controller) getNodeSubnet(netInfo util.NetInfo, nodeName string) ([]*ne
+	// Build ACLs once (not per switch - they're identical)
+	acls := c.buildPartialConnectivityACLs(cncName, hashNameV4, hashNameV6)
+
+	// Create/update ACLs - needed to be done only once - during updates this should be no-op
```
I think this comment from @jcaamano https://github.com/ovn-kubernetes/ovn-kubernetes/pull/5911#discussion_r2774970623 also answers this question. I think @npinaeva you were right that we always send an update if the entry is found and that we don't do anything smarter than that at libovsdbops level
cc @dave-tucker as well you were right too

tssurya commented:
The things I addressed in the recent push were:

1. https://github.com/ovn-kubernetes/ovn-kubernetes/pull/5828#issuecomment-3867316209 via commit https://github.com/ovn-kubernetes/ovn-kubernetes/pull/5828/changes/9920ef2bd799fee1f33004d570adf7b3b0f1c6e9
2. https://github.com/ovn-kubernetes/ovn-kubernetes/pull/5828#issuecomment-3867827435 via commit https://github.com/ovn-kubernetes/ovn-kubernetes/pull/5828/changes/0f33f0f52a66d5672cb1aa145225b360a7c478c2

I am working on addressing remaining review comments. I do think we probably need to squash the above two commits - but I didn't wanna make it hard for @npinaeva since I had to rebase so kept new change in commits along with why we redid the design with LBG - but I can also squash them into main commit.

tssurya commented on go-controller/pkg/ovn/controller/networkconnect/topology.go:
```diff
@@ -187,6 +220,19 @@ func (c *Controller) syncNetworkConnections(cnc *networkconnectv1.ClusterNetwork
+		// This creates the ACLs (idempotent) and adds them to the switch
+		if partialConnectivityDesired {
+			switchName, ok := partialConnState.networkSwitches[networkID]
+			if ok {
```
yup you are right, ok check is unnecessary
and switchname is also redundant as the function can just look it up

npinaeva commented on go-controller/pkg/ovn/controller/networkconnect/topology.go:
```diff
@@ -29,6 +29,28 @@ func getConnectRouterName(cncName string) string {
+// with its UUID populated. Returns nil, nil if the LBG doesn't exist.
+func (c *Controller) findCNCServiceLBGroup(cncName string) (*nbdb.LoadBalancerGroup, error) {
+	lbgName := getCNCServiceLBGroupName(cncName)
+	lbgs, err := libovsdbops.FindLoadBalancerGroupsWithPredicate(c.nbClient, func(lbg *nbdb.LoadBalancerGroup) bool {
```
Name is a server-side index for lb group, so we don't need a predicate

tssurya replied:
good point!

npinaeva commented on go-controller/pkg/ovn/controller/networkconnect/topology.go:
```diff
@@ -963,3 +1119,639 @@ func (c *Controller) getNodeSubnet(netInfo util.NetInfo, nodeName string) ([]*ne
+
+// ensureLoadBalancers finds and validates ClusterIP LBs for a single network.
+// Returns the found LBs and an error if the found count doesn't match the expected count.
+// This is a pure find+validate function -- it does NOT build any switch or LBG mutation ops.
```
I think the name should be changed then to `findLoadBalancers`, we usually use ensure with create db ops?

tssurya commented on go-controller/pkg/ovn/controller/networkconnect/topology.go:
```diff
@@ -963,3 +1119,639 @@ func (c *Controller) getNodeSubnet(netInfo util.NetInfo, nodeName string) ([]*ne
+
+	foundLBCount := len(thisnetworkLBs)
+
+	klog.V(5).Infof("Network %s (ID=%d): found %d ClusterIP LBs (expected %d)",
```
yea! exactly.. ideally it would be nice to add a "loadbalancer" create notifier that we can plug in.. but for now this works.. I'll add a comment to clarify that's why we are having to do this.
I hit this race in CI since we have e2e tests that check for "service didn't get any updates, all that happened was UDN got recreated" more details in https://github.com/ovn-kubernetes/ovn-kubernetes/pull/5828/changes/62f59792edeb0cdb115c9063cce194ee03f87912

tssurya commented:
all I did was plain rebase.. now working on addressing comments + remaining items in the PR describe... hold reviews..

tssurya commented:
> 3. The first commit which is a bug fix is not handling upgrades scenario - we probably at this point need to update the names of the load balancers at startup.
>    
>    * So looks like `EnsureLBs` is handling this correctly already. It takes `existingCacheLBs` and `LBs` desired.
>    * `toDelete` map initializes it with all existing cached LBs.
>    * `toDelete[lb.UUID] = lb` -- every existing LB starts as "to delete"
>    * `existingLB := existingByName[lb.Name]` -- matches by name since that's the index in DB
>    * `delete(toDelete, existingLB.UUID)` -- only removes from toDelete if name matches
>    * Everything still in toDelete gets deleted
>    * So: old per-node LB (name `_node_router+switch`) stays in `toDelete` because new desired LB has name `_cluster` (no name match). Old LB gets deleted
>    * I'll test and confirm this as well. THIS IS DONE, I checked it. We don't need to do anything specific for this.

proof of test:

```
I0211 10:52:56.570576   14254 services_controller.go:498] Services do not match for network=default, existing lbs: []services.LB{services.LB{Name:"green_green.network_Service_green/service-green_TCP_node_router+switch_ovn-control-plane", UUID:"82de3e7c-5df5-43f0-a257-35a0f62582a4", Protocol:"tcp", ExternalIDs:map[string]string{"k8s.ovn.org/kind":"Service", "k8s.ovn.org/network":"green_green-network", "k8s.ovn.org/owner":"green/service-green", "k8s.ovn.org/role":"primary"}, Opts:services.LBOpts{Reject:false, EmptyLBEvents:false, AffinityTimeOut:0, SkipSNAT:false, Template:false, AddressFamily:""}, Rules:[]services.LBRule{}, Templates:services.TemplateMap{}, Switches:[]string{"green_green.network_ovn-control-plane"}, Routers:[]string{"GR_green_green.network_ovn-control-plane"}, Groups:[]string{}}}, built lbs: []services.LB{services.LB{Name:"Service_green/service-green_TCP_cluster", UUID:"", Protocol:"TCP", ExternalIDs:map[string]string{"k8s.ovn.org/kind":"Service", "k8s.ovn.org/owner":"green/service-green"}, Opts:services.LBOpts{Reject:true, EmptyLBEvents:false, AffinityTimeOut:0, SkipSNAT:false, Template:false, AddressFamily:""}, Rules:[]services.LBRule{services.LBRule{Source:services.Addr{IP:"10.96.6.229", Port:80, Template:(*services.Template)(nil)}, Targets:[]services.Addr{services.Addr{IP:"10.244.1.4", Port:8080, Template:(*services.Template)(nil)}, services.Addr{IP:"10.244.2.7", Port:8080, Template:(*services.Template)(nil)}}}}, Templates:services.TemplateMap(nil), Switches:[]string{}, Routers:[]string{}, Groups:[]string{"clusterLBGroup"}}}
```
Before:
```
Defaulted container "nb-ovsdb" out of: nb-ovsdb, sb-ovsdb, ovn-northd, ovnkube-controller, ovn-controller, ovs-metrics-exporter
sh-5.2# ovn-nbctl find Load_Balancer | grep -E "name.*_node_router|name.*_cluster"
name                : "Service_frr-k8s-system/frr-k8s-metrics-service_TCP_node_router_ovn-control-plane"
name                : "Service_default/kubernetes_TCP_node_router_ovn-control-plane"
name                : "blue_blue.network_Service_blue/service-blue_TCP_node_router+switch_ovn-control-plane"
name                : "green_green.network_Service_green/service-green_TCP_node_router+switch_ovn-control-plane"
name                : "Service_kube-system/kube-dns_UDP_cluster"
name                : "Service_kube-system/kube-dns_TCP_cluster"
name                : "Service_frr-k8s-system/frr-k8s-webhook-service_TCP_node_router+switch_ovn-control-plane"
sh-5.2# ovn-nbctl list load-balancer blue_blue.network_Service_blue/service-blue_TCP_node_router+switch_ovn-control-plane
_uuid               : 3f286712-f760-4582-8a7d-1d3ae903beb3
external_ids        : {"k8s.ovn.org/kind"=Service, "k8s.ovn.org/network"=blue_blue-network, "k8s.ovn.org/owner"="blue/service-blue", "k8s.ovn.org/role"=primary}
health_check        : []
ip_port_mappings    : {}
name                : "blue_blue.network_Service_blue/service-blue_TCP_node_router+switch_ovn-control-plane"
options             : {event="false", hairpin_snat_ip="169.254.0.5 fd69::5", neighbor_responder=none, reject="true", skip_snat="false"}
protocol            : tcp
selection_fields    : []
vips                : {"10.96.124.96:80"="103.103.1.5:8080"}
sh-5.2# ovn-nbctl list load-balancer green_green.network_Service_green/service-green_TCP_node_router+switch_ovn-control-plane
_uuid               : 82de3e7c-5df5-43f0-a257-35a0f62582a4
external_ids        : {"k8s.ovn.org/kind"=Service, "k8s.ovn.org/network"=green_green-network, "k8s.ovn.org/owner"="green/service-green", "k8s.ovn.org/role"=primary}
health_check        : []
ip_port_mappings    : {}
name                : "green_green.network_Service_green/service-green_TCP_node_router+switch_ovn-control-plane"
options             : {event="false", hairpin_snat_ip="169.254.0.5 fd69::5", neighbor_responder=none, reject="true", skip_snat="false"}
protocol            : tcp
selection_fields    : []
vips                : {"10.96.6.229:80"="104.104.0.4:8080,104.104.1.4:8080"}
sh-5.2# 
sh-5.2# 
```
After:
```
Defaulted container "nb-ovsdb" out of: nb-ovsdb, sb-ovsdb, ovn-northd, ovnkube-controller, ovn-controller, ovs-metrics-exporter
sh-5.2# ovn-nbctl find Load_Balancer | grep -E "name.*_node_router|name.*_cluster"
name                : "Service_frr-k8s-system/frr-k8s-metrics-service_TCP_node_router_ovn-control-plane"
name                : "Service_default/kubernetes_TCP_node_router_ovn-control-plane"
name                : "Service_kube-system/kube-dns_UDP_cluster"
name                : "Service_kube-system/kube-dns_TCP_cluster"
name                : "Service_frr-k8s-system/frr-k8s-webhook-service_TCP_node_router+switch_ovn-control-plane"
name                : "Service_blue/service-blue_TCP_cluster"
name                : "Service_green/service-green_TCP_cluster"
sh-5.2#  ovn-nbctl list load-balancer Service_blue/service-blue_TCP_cluster
_uuid               : 66e2c7d6-f816-419d-b42d-4261041b772c
external_ids        : {"k8s.ovn.org/kind"=Service, "k8s.ovn.org/owner"="blue/service-blue"}
health_check        : []
ip_port_mappings    : {}
name                : "Service_blue/service-blue_TCP_cluster"
options             : {event="false", hairpin_snat_ip="169.254.169.5 fd69::5", neighbor_responder=none, reject="true", skip_snat="false"}
protocol            : tcp
selection_fields    : []
vips                : {"10.96.124.96:80"="10.244.1.3:8080"}
sh-5.2#  ovn-nbctl list load-balancer Service_green/service-green_TCP_cluster
_uuid               : 75f37727-10bd-40a3-9945-0458adcc5805
external_ids        : {"k8s.ovn.org/kind"=Service, "k8s.ovn.org/owner"="green/service-green"}
health_check        : []
ip_port_mappings    : {}
name                : "Service_green/service-green_TCP_cluster"
options             : {event="false", hairpin_snat_ip="169.254.169.5 fd69::5", neighbor_responder=none, reject="true", skip_snat="false"}
protocol            : tcp
selection_fields    : []
vips                : {"10.96.6.229:80"="10.244.1.4:8080,10.244.2.7:8080"}
```
resolving this.

tssurya commented on go-controller/pkg/ovn/controller/networkconnect/controller.go:
```diff
@@ -166,6 +202,7 @@ func (c *Controller) Start() error {
 		c.cncController,
 		c.nodeController,
 		c.nadReconciler,
+		c.serviceController,
```
this is done.

tssurya commented:
https://github.com/ovn-kubernetes/ovn-kubernetes/compare/ca6d7aa5ec7b28eda7a22a90b3f5b33de87baefb..690e7e0cc16a12309881cae66ca46f594e024c87 addresses

1. review comments and
2. adds more unit tests for service connectivity
3. and adds an initialSync
4. and fixes a bug around service update
 - all done in new commits on top, see commit messages for more information

I still need to add unit tests for partialconnectivity + rename things to `ServiceNetwork` - will do those two as new commits as well.

tssurya commented on go-controller/pkg/ovn/controller/networkconnect/controller.go:
```diff
@@ -332,6 +413,142 @@ func (c *Controller) reconcileNode(key string) error {
+
+	// Build owner key: "layer3_{networkID}" or "layer2_{networkID}"
+	return util.ComputeNetworkOwner(primaryNetwork.TopologyType(), primaryNetwork.GetNetworkID()), nil
+}
```
@coderabbitai: on thinking more.. but still i don't get one thing though.. when syncNAD requeues the CNC - shouldn't that take care of syncing the service as well? I don't think this is a bug is it?

tssurya commented on go-controller/pkg/ovn/controller/networkconnect/controller.go:
```diff
@@ -332,6 +417,144 @@ func (c *Controller) reconcileNode(key string) error {
+	if err != nil && !apierrors.IsNotFound(err) {
+		return err
+	}
+	// Service may be nil if deleted (this shouldn't happen since we only process create events)
```
yea :D oops fixing this

tssurya commented:
I'll rebase on master since we merged https://github.com/ovn-kubernetes/ovn-kubernetes/pull/5886  just to be in the clear of no failures..
no functional changes. only the last 3 commits are the new ones compared to our last rounds..

tssurya commented:
https://github.com/ovn-kubernetes/ovn-kubernetes/compare/d5843bd973aa7b88bc75be9118c85fd113cb87d7..4a1797d30df78e8e706a7aeb0a3efc4efdb557cd

is the diff

I'll also now rebase on master while at it.
